iceberg/iceberg-catalog/src/main/java/org/apache/iceberg/hive/HiveSchemaUtil.java

```diff
@@ -138,22 +137,88 @@ public static Type convert(TypeInfo typeInfo) {
   }
 
   /**
-   * Produces the difference of two FieldSchema lists by only taking into account the field name and type.
+   * Returns a SchemaDifference containing those fields which are present in only one of the collections, as well as
+   * those fields which are present in both (in terms of the name) but their type or comment has changed.
    * @param minuendCollection Collection of fields to subtract from
    * @param subtrahendCollection Collection of fields to subtract
-   * @return the result list of difference
+   * @param bothDirections Whether or not to compute the missing fields from the minuendCollection as well
+   * @return the difference between the two schemas
    */
-  public static Collection<FieldSchema> schemaDifference(
-      Collection<FieldSchema> minuendCollection, Collection<FieldSchema> subtrahendCollection) {
+  public static SchemaDifference getSchemaDiff(Collection<FieldSchema> minuendCollection,
+                                               Collection<FieldSchema> subtrahendCollection, boolean bothDirections) {
+    SchemaDifference difference = new SchemaDifference();
 
-    Function<FieldSchema, FieldSchema> unsetCommentFunc = fs -> new FieldSchema(fs.getName(), fs.getType(), null);
-    Set<FieldSchema> subtrahendWithoutComment =
-        subtrahendCollection.stream().map(unsetCommentFunc).collect(Collectors.toSet());
+    for (FieldSchema first : minuendCollection) {
+      boolean found = false;
+      for (FieldSchema second : subtrahendCollection) {
+        if (first.getName().equals(second.getName())) {
```

Comment:
Why not use Objects.equals here too?

Reply:
I didn't use it because `name` should never be null, unlike comment (and maybe type?). Anyways, I've changed it to use Objects.equals to be safe and for symmetry.